### PR TITLE
Checksum::Store mutex

### PR DESCRIPTION
**Pull request by martinemde**
## What was the end-user or developer problem that led to this PR?

Although ruby/ruby#9240 may fix the underlying problem, the hash access errors in Checksum::Store are cause for some concerns.

## What is your fix for the problem, implemented in this PR?

Use a mutex to synchronize checksum store hash access.

## Make sure the following tasks are checked

- [X] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [X] Write code to solve the problem
- [X] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/lib/bundler/checksum.rb [hunk 1]
```diff
@@ -161,26 +165,17 @@ class Store
 
       def initialize
         @store = {}
-      end
-
-      def initialize_copy(other)
```
Comment (segiddins) [inline bundler/lib/bundler/checksum.rb:166]: Why are you removing this?

File: bundler/lib/bundler/checksum.rb [hunk 2]
```diff
@@ -228,16 +223,32 @@ def to_lock(spec)
...
+        if existing
+          merge_checksum(lock_name, checksum, existing)
         else
-          raise ChecksumMismatchError.new(lock_name, existing, checksum)
+          store_checksum(lock_name, checksum)
+        end
+      end
+
+      def merge_checksum(lock_name, checksum, existing)
+        existing.merge!(checksum) || raise(ChecksumMismatchError.new(lock_name, existing, checksum))
```
Comment (segiddins) [inline bundler/lib/bundler/checksum.rb:236]: Should this also be protected by the mutex?

File: bundler/lib/bundler/checksum.rb (hunk 1 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:166]: It's never used, but if you think there's value to it. I couldn't think of a case where we'd dup it legitimately and there's no test for it.

File: bundler/lib/bundler/checksum.rb (hunk 2 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:236]: `merge!` just compares the digest/algo, concats the `source` onto `sources`, then runs `sources.uniq!`. Even run concurrently without global lock, I think that's still safe.

File: bundler/lib/bundler/checksum.rb (hunk 1 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:166]: In actuality, when I was adding the mutex I had to duplicate the mutex into here, then also lock the copy process and it seemed excessive for something that we don't use.

Comment (segiddins) [reply bundler/lib/bundler/checksum.rb:166]: ah if it is never used then we're good, I remember it was used at some point

File: bundler/lib/bundler/checksum.rb (hunk 2 above)
Comment (segiddins) [reply bundler/lib/bundler/checksum.rb:236]: if it's modifying any shared state, it is not safe -- remember not all Rubies have a GIL

Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:236]: How can you tell which operations are safe on global state, and which aren't? This is tue code:

```
@sources.concat(other.sources).uniq!
```

This brings up another situation in here that isn't locked. I'll add a comment there.

File: bundler/lib/bundler/checksum.rb [hunk 3]
```diff
@@ -228,16 +223,32 @@ def to_lock(spec)
...
+      end
+
+      def store_checksum(lock_name, checksum)
+        @store_mutex.synchronize do
+          (@store[lock_name] ||= {}).store(checksum.algo, checksum)
+        end
+      end
+
+      def fetch_checksum(lock_name, algo)
+        fetch_checksums(lock_name)&.fetch(algo, nil)
```
Comment (martinemde) [inline bundler/lib/bundler/checksum.rb:246]: @segiddins This is a fetch on a Hash (the nested one). We're syncing other reads, but is this one safe? Note the second half of this method is not synced.

Comment (segiddins) [reply bundler/lib/bundler/checksum.rb:246]: any read or write on data that is (at the time of access) shared across threads (or fibers, actors, etc), needs to be synchronized. typically, this means not building upon primitive access methods directly, but rather exposing a set of public methods that each synchronizes access to the shared data

**Review by segiddins — APPROVED** [review]

File: bundler/lib/bundler/checksum.rb [hunk 4]
```diff
@@ -217,18 +222,27 @@ def to_lock(spec)
       private
 
       def register_checksum(lock_name, checksum)
-        return unless checksum
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/checksum.rb:231]: You realized we were always passing non nil here, correct?

File: bundler/lib/bundler/checksum.rb [hunk 5]
```diff
@@ -179,16 +185,14 @@ def inspect
...
-        existing = checksums[checksum.algo]
-
-        # we assume only one source because this is used while building the index
-        if !existing || existing.sources.first == checksum.sources.first
-          checksums[checksum.algo] = checksum
-        else
-          register_checksum(lock_name, checksum)
+        @store_mutex.synchronize do
+          existing = fetch_checksum(lock_name, checksum.algo)
+          if !existing || existing.same_source?(checksum)
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/checksum.rb:191]: The implementation of this method sounds slightly different than before? Also should the comment be carried over?

File: bundler/lib/bundler/checksum.rb [hunk 6]
```diff
@@ -179,16 +185,14 @@ def inspect
       # that contain the same gem with different checksums.
       def replace(spec, checksum)
         return unless checksum
-
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/checksum.rb:193]: We do generally keep a blank line after guards, I think.

**Review by deivid-rodriguez — COMMENTED** [review]
Looks very good, just some minor comments on stuff unrelated to the "mutexing".

File: bundler/lib/bundler/checksum.rb (hunk 4 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:231]: Yes, this was always guarded by the callers, and the method is private.

File: bundler/lib/bundler/checksum.rb (hunk 5 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:191]: I short-circuited. `register_checksum` was re-fetching the existing checksum when we already have it here.

File: bundler/lib/bundler/checksum.rb (hunk 6 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:193]: Yeah, this is a bit claustrophobic. Will fix.

File: bundler/lib/bundler/checksum.rb (hunk 5 above)
Comment (martinemde) [reply bundler/lib/bundler/checksum.rb:191]: The comment is no longer correct. I made it so same_source checks for any matching source on existing, and the incoming one should always have only 1. I think the method name explains it well enough now.

Comment (deivid-rodriguez) [reply bundler/lib/bundler/checksum.rb:191]: I meant the one you extracted, `same_source?`.

Comment (deivid-rodriguez) [reply bundler/lib/bundler/checksum.rb:196]: I get it now!

**Pull request merged**